Summarizing chat memory

PaperFly-web opened the pull request:
## Change
Added `SummarizingChatMemory` class to support automatic summarization of chat history.

Features include:
- Automatic summarization when message count exceeds a configurable threshold
- Support for both static and dynamic `maxMessages` and `summarizeThreshold`
- Dynamic system prompt generation per memory ID for context-aware summaries
- Automatic cleanup of orphan `ToolExecutionResultMessage` when associated `AiMessage` is evicted
- Ensures `SystemMessage` is always preserved at the head of memory
- Works with any `ChatMemoryStore` for persistence
- Builder API with clear and flexible configuration
- Comprehensive unit tests covering edge cases and dynamic configuration

Example usage:

```java
ChatMemory chatMemory = SummarizingChatMemory.builder()
    .chatModel(chatModel)
    .maxMessages(4)
    .summarizeThreshold(2)
    .dynamicSystemPrompt(id -> "Dynamic prompt for memory " + id)
    .build();
```



## General checklist
<!-- Please double-check the following points and mark them like this: [X] -->
- [X] There are no breaking changes (API, behaviour)
- [X] I have added unit and/or integration tests for my change
- [X] The tests cover both positive and negative cases
- [X] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [X] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
<!-- Before adding documentation and example(s) (below), please wait until the PR is reviewed and approved. -->
- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)


## Checklist for adding new maven module
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`


## Checklist for adding new embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

PaperFly-web committed:
feat(memory): Add summarization-based chat memory supporting automatic conversation compression

- Implemented SummarizingChatMemory class for managing conversation history
- Automatically triggers summarization when the number of messages exceeds a defined threshold
- Supports dynamic injection of maxMessages and summarizeThreshold via function references
- Allows generating dynamic system prompts for each memory ID
- Checks and performs summarization logic when adding or retrieving messages
- Provides a flexible Builder pattern for configuration
- Automatically removes duplicate system messages and handles tool execution result messages
- Generates conversation summaries using the specified system prompt and chat model

PaperFly-web committed:
feat(memory): Implement chat memory with automatic summarization capability

Introduce the SummarizingChatMemory class, which automatically summarizes conversation history when the message count exceeds a configured threshold.
Support both static and dynamic configuration of maximum message count (maxMessages) and summarization threshold (summarizeThreshold).
Enable dynamic generation of system prompts per memory ID to provide context-aware summarization instructions.
Automatically clean up orphaned ToolExecutionResultMessage instances when their associated AiMessage is removed.
Ensure that SystemMessage is always retained at the beginning of the memory and is never summarized or removed.
Integrate with any ChatMemoryStore to persist memory state.
Provide a comprehensive builder API and usage examples to enhance configurability and usability.
Add extensive unit tests covering various edge cases and dynamic configuration scenarios.

PaperFly-web committed:
test (memory): Update the AI message content in the test case
-Modified the content of the ai2 message in Summarizing Chat Memory Test and added an exclamation mark to enhance test coverage

PaperFly-web committed:
test (memory): Update OpenAI model configuration in SummarizingChatMemory Test
-Change baseURL from fixed URL to environment variable OPENAI_SASE-URL
-Change apiKey from deepseed key to OPENAI_SPI_KEY - Add organizationId configuration item
-Change the modelName from deepseek chat to GPT4_onMINI
-Keep logRequests and logResponses configurations
-Add TODO annotation plan to support more models

PaperFly-web committed:
docs (memory): Update JavaDoc comments for SummarizingChatMemory class
-Removed author annotations at the class level and moved them before the class definition
-Simplified title format for key features and usage instructions
-Deleted redundant design description paragraphs - fixed formatting issues in code examples
-Removed explicit comments from default system prompts - removed delimiter comments for constructors, public APIs, and internal logic - fixed space formatting issues around variable names
-Removed the separation comments in the builder section

PaperFly-web committed:
refactor(memory): Formats code spotlessly

PaperFly-web committed:
Merge branch 'main' into summarizing-chat-memory

PaperFly-web committed:
fix(memory): Fix incorrect reference to system prompt function - Correct the wrong "chatModel" reference to "systemPromptFunction"
- Ensure that the system prompts the function to initialize and use correctly
- Fixed an issue where the null pointer could be abnormally caused by this

dliubarskyi commented:
cc @cescoffier you might be interested in this PR

dliubarskyi commented on langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java:
```diff
@@ -0,0 +1,340 @@
+    /**
+     * Generate a concise AI summary for the removed messages.
+     */
+    private AiMessage generateSummary(List<ChatMessage> messages) {
```
Not sure that summary should reside in the `AiMessage`. Perhaps better to keep it in the `UserMessage`?

PaperFly-web replied:
Yes, you're absolutely right — keeping the summary in the UserMessage makes more sense. I’ll update it to use UserMessage instead.

dliubarskyi commented on langchain4j/src/test/java/dev/langchain4j/memory/chat/SummarizingChatMemoryTest.java:
```diff
@@ -0,0 +1,782 @@
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+@EnabledIfEnvironmentVariable(named = "deepseek-key", matches = ".+")
```
```suggestion
@EnabledIfEnvironmentVariable(named = "OPENAI_API_KEY", matches = ".+")
```

PaperFly-web replied:
OK, I will replace it.

dliubarskyi commented on langchain4j/src/test/java/dev/langchain4j/memory/chat/SummarizingChatMemoryTest.java:
```diff
@@ -0,0 +1,782 @@
+
+        // Verify that the first three messages are still retained
+        assertThat(chatMemory.messages())
+                .usingRecursiveFieldByFieldElementComparatorIgnoringFields("id", "timestamp", "createdAt")
```
why is this required?

dliubarskyi commented on langchain4j/src/test/java/dev/langchain4j/memory/chat/SummarizingChatMemoryTest.java:
```diff
@@ -0,0 +1,782 @@
+                .usingRecursiveFieldByFieldElementComparatorIgnoringFields("id", "timestamp", "createdAt")
+                .containsExactly(userMessage("hello"), userMessage("world"), userMessage("I am lisi"));
+
+        // Add the fourth message, which may trigger summarization logic
```
why this comment says that it "may" trigger? This test should be fully deterministic

PaperFly-web replied:
This is my mistake — I didn’t check this comment carefully. I’ll review it again to make sure it’s correct.This probably won’t be triggered now — it will only be triggered the next time something is added.

dliubarskyi commented on langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java:
```diff
@@ -0,0 +1,340 @@
+     */
+    private AiMessage generateSummary(List<ChatMessage> messages) {
+        SystemMessage systemPrompt = SystemMessage.from(systemPromptFunction.apply(id));
+        messages.add(0, systemPrompt);
```
this seems odd, why do you insert the summarization message in the memory?

dliubarskyi replied:
my bad - it is not memory, but removed messages. The name of the variable is confusing. In any case it is not the best practice to modify collections passed into method as arguments

PaperFly-web replied:
Okay, create a new copy of the list to avoid modifying the passed parameters.

dliubarskyi commented on langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java:
```diff
@@ -0,0 +1,340 @@
+
+/**
+ * A chat memory implementation that automatically summarizes conversation history
+ * when the number of messages exceeds a configurable threshold.
```
```suggestion
 * when the number of messages exceeds a configurable `maxMessages`.
```
Otherwise it can be confused with `summarizeThreshold`

dliubarskyi replied:
But in general I think it would be more useful to have `maxTokens` setting instead of `maxMessages`, as it is much more precise and controllable

PaperFly-web replied:
Indeed, maxTokens would be better. It can prevent the summary from exceeding the token limit of the large model due to an excessive number of messages. I will switch to using tokens to summarize the conversation.

PaperFly-web replied:
However, if we switch to using maxTokens, it becomes difficult to control whether the generated summary exceeds maxTokens. Therefore, I think we can consider providing an interface function, allowing users to summarize according to their own strategies if they have any.

dliubarskyi commented on langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java:
```diff
@@ -0,0 +1,340 @@
+ * <ul>
+ *     <li>Supports both static and dynamic configuration of <b>maxMessages</b>:
+ *         the maximum number of messages retained before triggering evictions.</li>
+ *     <li>Supports both static and dynamic configuration of <b>summarizeThreshold</b>:
```
I am not sure `summarizeThreshold` is the clear enough name, perhaps something like `maxMessagesToSummarize`?

PaperFly-web replied:
Thank you for your suggestion. Indeed, "maxMessagesToSummarize" has a better semantic meaning, and I will replace it.

PaperFly-web committed:
refactor(memory): Refactor SummarizingChatMemory to support custom summary generation functions

- Renamed summarizeThreshold to maxMessagesToSummarize for clearer semantics.
- Removed direct dependencies on ChatModel and systemPromptFunction.
- Introduced generateSummaryFunction to enable custom summarization logic.
- Changed default summarization function to return a UserMessage instead of an AiMessage.
- Updated all related test cases to align with the new configuration approach.
- Revised Javadoc comments to reflect the new parameter names and behavior.
- Deleted deprecated methods dynamicSystemPrompt and systemPrompt.
- Simplified internal message insertion logic, unifying it under the summary function.

PaperFly-web committed:
test(memory): Optimize the chat memory test code

- Introduced the OpenAiChatModelName constant to improve model name readability
- Use variables to reference user messages to avoid creating objects repeatedly
- Simplified assertion logic, removal of redundant field ignore configurations
- Updated OpenAI model configuration to enable organization ID and specify gpt-4o-mini model - Removed no-note comments to improve code cleanliness

PaperFly-web committed:
test(memory): Optimized the test logic for summarizing chat memory - Adjusted the order of message addition to verify summary behavior more accurately
- Remove redundant assertion comparison configurations to simplify test code
- Updated parameter naming to improve readability (summarizeThreshold → maxMessagesToSummarize)
- Cleaned up outdated model configuration references in comments
- Simplified logical representation of the dynamic parameter update section
- Unify variable naming styles in test cases
- Remove unnecessary empty lines and comment blocks
- Updated the model name to "deepseek-chat" and commented out the organization ID configuration

dliubarskyi commented:
BTW there was an old PR that tried to do a similar thing, might be worth checking it for ideas: https://github.com/langchain4j/langchain4j/pull/1351

PaperFly-web commented:
> BTW there was an old PR that tried to do a similar thing, might be worth checking it for ideas: #1351

Thank you for your reminder. I checked it out and feel that his version is a bit messy. He generated system messages through summarization. Logically speaking, system messages should not be modified casually; generally, it's better to generate user messages. However, I still appreciate your reminder. I will carefully look at his code again to gain some inspiration.

PaperFly-web committed:
fix(memory): remove redundant language requirement from summarization prompt

- Removed unnecessary constraint specifying Chinese as output language
- Maintained existing requirements for summary length and content accuracy
- Kept exclusion rule for irrelevant dialogue and system messages
- Preserved natural and fluent writing requirement
- Simplified prompt by removing conflicting language specification
- Ensured summary generation remains focused on main topics and conclusions

douglas-DS commented:
@PaperFly-web @dliubarskyi  I’m wondering whether using a _maxMessages_ threshold is the best approach here. Most AI frameworks I’ve seen apply summarization based on a token window, typically controlled by a _maxTokens_ setting. This ensures the memory is kept within the configured context size. Would it make more sense to follow that pattern here as well?

dliubarskyi commented:
@douglas-DS [agreed](https://github.com/langchain4j/langchain4j/pull/4003#discussion_r2519059833)

PaperFly-web commented:
> @PaperFly-web @dliubarskyi I’m wondering whether using a _maxMessages_ threshold is the best approach here. Most AI frameworks I’ve seen apply summarization based on a token window, typically controlled by a _maxTokens_ setting. This ensures the memory is kept within the configured context size. Would it make more sense to follow that pattern here as well?
@douglas-DS 
Sorry, I've been busy with work lately and forgot to reply. I've also tried using token calculation, but the final summary generated by the large language model cannot be controlled to a specific number of tokens. So, you might have defined summarizing once with more than 1000 tokens, but this summary might generate more than 1000 tokens of data. This would cause a conflict between the summary and the user-defined token threshold. Therefore, I don't have a better solution to this problem at the moment.

dliubarskyi commented:
Thanks a lot for your PR! Unfortunately, we have to close it. See more details here: https://github.com/langchain4j/langchain4j/issues/5443#issuecomment-4707891571
